executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+                    ));
+                }
+                Effect::RangeConstraint(..) => {
+                    unreachable!("Final code should not contain pure range constraints.")
+                }
+                Effect::Assertion(Assertion {
+                    lhs,
+                    rhs,
+                    expected_equal,
+                }) => {
+                    actions.push(InterpreterAction::Assertion(
+                        lhs.map_variables(&mut map_var_idx).to_rpn(),
+                        rhs.map_variables(&mut map_var_idx).to_rpn(),
+                        *expected_equal,
+                    ));
+                }
+                Effect::MachineCall(id, arguments) => {
+                    let mut result_vars = vec![];
+
+                    arguments.iter().for_each(|a| match a {
+                        MachineCallArgument::Unknown(v) => {
+                            let idx = map_var_idx(v);
+                            result_vars.push(idx);
+                        }
+                        MachineCallArgument::Known(_) => {}
+                    });
+
+                    actions.push(InterpreterAction::MachineCall(
+                        result_vars,
+                        *id,
+                        arguments
+                            .iter()
+                            .map(|a| match a {
+                                MachineCallArgument::Unknown(_) => MachineCallArgument::Unknown(0),
+                                MachineCallArgument::Known(v) => {
+                                    MachineCallArgument::Known(v.map_variables(&mut map_var_idx))
+                                }
+                            })
+                            .collect(),
+                    ));
+                }
+            }
+        }
+
+        // write variables back
+        let vars_known: Vec<_> = effects.iter().flat_map(written_vars_in_effect).collect();
+        vars_known.iter().for_each(|var| {
+            match var {
+                Variable::Cell(cell) => {
+                    let idx = var_idx.get(var).unwrap();
+                    actions.push(InterpreterAction::WriteCell(*idx, cell.clone()));
+                    actions.push(InterpreterAction::WriteKnown(cell.clone()));
+                }
+                Variable::Param(i) => {
+                    let idx = var_idx.get(var).unwrap();
+                    actions.push(InterpreterAction::WriteParam(*idx, *i));
+                }
+                Variable::MachineCallReturnValue(_) => {
+                    // This is just an internal variable.
+                }
+            }
+        });
+
+        Self {
+            first_column_id,
+            column_count,
+            vars,
+            actions,
+        }
+    }
+
+    // Execute the machine effects for the given the parameters
+    pub fn call<Q: QueryCallback<T>>(&self, params: WitgenFunctionParams<'_, T>) {
```

Comment:
Is it possible to use a higher-level interface than WitgenFunctionParams? They require unsafe code to be used and are tailored towards an FFI.

Reply:
I think `CompactDataRef<'_, T>,` would be the right thing to pass